### ISSUE TITLE
Numerical overflow / instability with Poisson loss

### ISSUE THREAD

**Issue by bduclaux** [catboost/catboost#1829]
Problem: Poisson loss should better manage numerical overflows

catboost version: 0.26.1
Operating System: Linux - CPU training

Consider the follow dataset columns:
```
0	Label
1	Categ	f1
2	Num	f2
```

and  the following artificial dataset with 10 records:
```
100	1	15
1	2	55
1	3	120
1	4	33
0	5	150
0	6	120
0	7	900
0	8	230
0	9	60
0	10	140
```

Fit the model with 100 iterations, with default parameters:
```
catboost fit -i 100 --learn-set data --cd cols.cd --loss-function Poisson 
link with lfalloc for better performance
0:	learn: -6.3245933	total: 1.54ms	remaining: 152ms
1:	learn: -11.9244648	total: 1.89ms	remaining: 92.7ms
2:	learn: -16.0195799	total: 2.05ms	remaining: 66.3ms
3:	learn: -18.7436854	total: 2.23ms	remaining: 53.5ms
98:	learn: -35.4243363	total: 17.2ms	remaining: 173us
99:	learn: -35.4264662	total: 17.3ms	remaining: 0us
Skipping test eval output
0.00167070143 min passed
```

So far, so good. Predictions are almost right:
```
catboost calc --input-path data --output-columns Label,Exponent
Label	Exponent
100	94.01656361
1	0.5724194454
1	0.4529287883
1	0.8418728166
0	0.4296197358
0	0.4529287883
0	0.434384209
0	0.4294073225
0	0.4898972607
0	0.4403833727
```

Now change the label of the first record from 100 to 10000:
`10000	1	15`

And fit again:
```
catboost fit -i 100 --learn-set data --cd cols.cd --loss-function Poisson 
link with lfalloc for better performance
0:	learn: 37053412063057383598061451739136.0000000	total: 1.84ms	remaining: 182ms
1:	learn: 35958318263438618937321572007936.0000000	total: 2.02ms	remaining: 99.1ms
2:	learn: 34895589376069090663726869118976.0000000	total: 2.69ms	remaining: 87.1ms
3:	learn: 33864268873256524021807060090880.0000000	total: 2.88ms	remaining: 69.2ms
....
99:	learn: 1900962824122301108392223047680.0000000	total: 17ms	remaining: 0us
Skipping test eval output
0.001649521296 min passed

```

We have now a nice overflow during training.

Predictions are now completely off:
```
catboost calc --input-path data --output-columns Label,Exponent
Label	Exponent
10000	1.900962824e+31
1	0.4221586681
1	0.3818648805
1	0.4228914735
0	0.3571887462
```

It is likely that the loss calculation exp(ai) - ai . ti , documented at https://catboost.ai/docs/concepts/loss-functions-regression.html results in an overflow when there are some extreme value ranges in the dataset (here from 1 to 10'000).
Also, with some other datasets, Poisson loss suddently explodes in the middle of training. I can provide such a dataset privately. 
The approach we use to cope with this problem is winsorization, but this is clearly not optimal. 
I hope the Catboost team can implement a fix to this problem!

**Comment by bduclaux** [comment]
Hello, 

In addition to the previous issue, here is a trace below which show another loss instability (or issue) while training a Poisson regression model:

```
103:    learn: 0.1950309        total: 30.3s    remaining: 1m 55s
104:    learn: 0.1949870        total: 30.6s    remaining: 1m 55s
105:    learn: 0.1948367        total: 30.9s    remaining: 1m 54s
106:    learn: 0.1939248        total: 31.2s    remaining: 1m 54s
107:    learn: 0.1938864        total: 31.5s    remaining: 1m 54s
108:    learn: 0.1935502        total: 31.8s    remaining: 1m 54s
109:    learn: 0.1935353        total: 32.1s    remaining: 1m 53s
110:    learn: 0.1934567        total: 32.3s    remaining: 1m 53s
111:    learn: 0.1933285        total: 32.6s    remaining: 1m 52s
112:    learn: 4.2927961        total: 33s      remaining: 1m 53s
113:    learn: 3.5829970        total: 33.4s    remaining: 1m 53s
114:    learn: 3.0024428        total: 33.7s    remaining: 1m 52s
115:    learn: 2.5238969        total: 34s      remaining: 1m 52s
116:    learn: 2.1202193        total: 34.4s    remaining: 1m 52s
117:    learn: 1.7863100        total: 34.6s    remaining: 1m 52s
118:    learn: 1.5187816        total: 35s      remaining: 1m 51s
119:    learn: 1.3027842        total: 35.2s    remaining: 1m 51s
120:    learn: 1.1095979        total: 35.4s    remaining: 1m 50s
121:    learn: 0.9617626        total: 35.8s    remaining: 1m 50s
122:    learn: 0.8276232        total: 36s      remaining: 1m 50s
123:    learn: 0.7175082        total: 36.2s    remaining: 1m 49s
124:    learn: 0.6376748        total: 36.6s    remaining: 1m 49s
125:    learn: 0.5691369        total: 37s      remaining: 1m 49s
126:    learn: 0.5032073        total: 37.2s    remaining: 1m 49s
127:    learn: 0.4533072        total: 37.6s    remaining: 1m 49s
```

As you can see, the loss starts to explode at iteration 112. 
Looks like Catboost has implemented a Cyclical learning rate :-)
Joking aside, it would be useful to understand the reasons for such problem.
The Poisson model was trained with the following parameters:
` -i 500 -n 6 -w 0.19 --task-type CPU --max-ctr-complexity 2 --leaf-estimation-iterations 1`

Hope it helps.

**Comment by bduclaux** [comment]
Hello,

Here is another strange case, where the Poisson loss suddenly starts to grow, but not in large proportions (see iteration 410), then decreases again:

```
400:    learn: 0.1516693        total: 19m 1s   remaining: 4m 41s
401:    learn: 0.1516552        total: 19m 6s   remaining: 4m 39s
402:    learn: 0.1516455        total: 19m 8s   remaining: 4m 36s
403:    learn: 0.1516338        total: 19m 12s  remaining: 4m 33s
404:    learn: 0.1516335        total: 19m 14s  remaining: 4m 30s
405:    learn: 0.1516289        total: 19m 17s  remaining: 4m 28s
406:    learn: 0.1515938        total: 19m 21s  remaining: 4m 25s
407:    learn: 0.1515890        total: 19m 24s  remaining: 4m 22s
408:    learn: 0.1515831        total: 19m 27s  remaining: 4m 19s
409:    learn: 0.1515700        total: 19m 31s  remaining: 4m 17s
410:    learn: 0.1670750        total: 19m 34s  remaining: 4m 14s
411:    learn: 0.1659063        total: 19m 37s  remaining: 4m 11s
412:    learn: 0.1633739        total: 19m 41s  remaining: 4m 8s
413:    learn: 0.1628806        total: 19m 45s  remaining: 4m 6s
414:    learn: 0.1621478        total: 19m 47s  remaining: 4m 3s
415:    learn: 0.1618619        total: 19m 51s  remaining: 4m
416:    learn: 0.1616720        total: 19m 54s  remaining: 3m 57s
417:    learn: 0.1598617        total: 19m 58s  remaining: 3m 55s
418:    learn: 0.1597656        total: 20m 1s   remaining: 3m 52s
419:    learn: 0.1597029        total: 20m 4s   remaining: 3m 49s
420:    learn: 0.1594270        total: 20m 7s   remaining: 3m 46s
421:    learn: 0.1593618        total: 20m 10s  remaining: 3m 43s
422:    learn: 0.1589985        total: 20m 13s  remaining: 3m 40s
423:    learn: 0.1589465        total: 20m 16s  remaining: 3m 38s
```

**Comment by Evgueni-Petrov-aka-espetrov** [comment]
@bduclaux many thanks for the reproducer! Will be looking for a solution.

**Comment by Evgueni-Petrov-aka-espetrov** [comment]
Reproducer
```python
import catboost

# Initialize data
cat_features = [1]
train_data = [
    [1, 15],
    [2, 55],
    [3, 120],
    [4, 33],
    [5, 150],
    [6, 120],
    [7, 900],
    [8, 230],
    [9, 60],
    [10, 140]
]
train_labels = [100, 1, 1, 1, 0, 0, 0, 0, 0, 0]

# Initialize CatBoostClassifier
model = catboost.CatBoostRegressor(iterations=20, loss_function='Poisson')
# Fit model
model.fit(train_data, train_labels, cat_features=cat_features)
```
